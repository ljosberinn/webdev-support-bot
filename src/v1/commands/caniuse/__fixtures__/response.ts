export const simpleResponse = 'flexbox-gap,flexbox';
export const detailedResponse = [
  {
    baseCategories: ['CSS'],
    bug_count: 0,
    children: null,
    chrome_id: '',
    date_published: '2019-06-17',
    description:
      '`gap` for flexbox containers to create gaps/gutters between flex items',
    firefox_id: '',
    ie_id: '',
    keywords: 'flexbox,gaps,gutters,row-gap,column-gap',
    link_count: 5,
    notes: '',
    notes_by_num: {
      '1':
        'Can be enabled in Chrome via the `#enable-experimental-web-platform-features` flag in `chrome://flags`',
    },
    parent: 'flexbox',
    spec: 'https://www.w3.org/TR/css-align-3/#gaps',
    stats: {
      and_chr: {
        '81': 'n',
      },
      and_ff: {
        '68': 'y',
      },
      and_qq: {
        '10.4': 'n',
      },
      and_uc: {
        '12.12': 'n',
      },
      android: {
        '2.1': 'n',
        '2.2': 'n',
        '2.3': 'n',
        '3': 'n',
        '4': 'n',
        '4.1': 'n',
        '4.2-4.3': 'n',
        '4.4': 'n',
        '4.4.3-4.4.4': 'n',
        '81': 'n',
      },
      baidu: {
        '7.12': 'n',
      },
      bb: {
        '10': 'n',
        '7': 'n',
      },
      chrome: {
        '10': 'n',
        '11': 'n',
        '12': 'n',
        '13': 'n',
        '14': 'n',
        '15': 'n',
        '16': 'n',
        '17': 'n',
        '18': 'n',
        '19': 'n',
        '20': 'n',
        '21': 'n',
        '22': 'n',
        '23': 'n',
        '24': 'n',
        '25': 'n',
        '26': 'n',
        '27': 'n',
        '28': 'n',
        '29': 'n',
        '30': 'n',
        '31': 'n',
        '32': 'n',
        '33': 'n',
        '34': 'n',
        '35': 'n',
        '36': 'n',
        '37': 'n',
        '38': 'n',
        '39': 'n',
        '4': 'n',
        '40': 'n',
        '41': 'n',
        '42': 'n',
        '43': 'n',
        '44': 'n',
        '45': 'n',
        '46': 'n',
        '47': 'n',
        '48': 'n',
        '49': 'n',
        '5': 'n',
        '50': 'n',
        '51': 'n',
        '52': 'n',
        '53': 'n',
        '54': 'n',
        '55': 'n',
        '56': 'n',
        '57': 'n',
        '58': 'n',
        '59': 'n',
        '6': 'n',
        '60': 'n',
        '61': 'n',
        '62': 'n',
        '63': 'n',
        '64': 'n',
        '65': 'n',
        '66': 'n',
        '67': 'n',
        '68': 'n',
        '69': 'n',
        '7': 'n',
        '70': 'n',
        '71': 'n',
        '72': 'n',
        '73': 'n',
        '74': 'n',
        '75': 'n',
        '76': 'n',
        '77': 'n',
        '78': 'n',
        '79': 'n',
        '8': 'n',
        '80': 'n',
        '81': 'n',
        '83': 'n',
        '84': 'y',
        '85': 'y',
        '86': 'y',
        '9': 'n',
      },
      edge: {
        '12': 'n',
        '13': 'n',
        '14': 'n',
        '15': 'n',
        '16': 'n',
        '17': 'n',
        '18': 'n',
        '79': 'n',
        '80': 'n',
        '81': 'n',
        '83': 'n',
      },
      firefox: {
        '10': 'n',
        '11': 'n',
        '12': 'n',
        '13': 'n',
        '14': 'n',
        '15': 'n',
        '16': 'n',
        '17': 'n',
        '18': 'n',
        '19': 'n',
        '2': 'n',
        '20': 'n',
        '21': 'n',
        '22': 'n',
        '23': 'n',
        '24': 'n',
        '25': 'n',
        '26': 'n',
        '27': 'n',
        '28': 'n',
        '29': 'n',
        '3': 'n',
        '3.5': 'n',
        '3.6': 'n',
        '30': 'n',
        '31': 'n',
        '32': 'n',
        '33': 'n',
        '34': 'n',
        '35': 'n',
        '36': 'n',
        '37': 'n',
        '38': 'n',
        '39': 'n',
        '4': 'n',
        '40': 'n',
        '41': 'n',
        '42': 'n',
        '43': 'n',
        '44': 'n',
        '45': 'n',
        '46': 'n',
        '47': 'n',
        '48': 'n',
        '49': 'n',
        '5': 'n',
        '50': 'n',
        '51': 'n',
        '52': 'n',
        '53': 'n',
        '54': 'n',
        '55': 'n',
        '56': 'n',
        '57': 'n',
        '58': 'n',
        '59': 'n',
        '6': 'n',
        '60': 'n',
        '61': 'n',
        '62': 'n',
        '63': 'y',
        '64': 'y',
        '65': 'y',
        '66': 'y',
        '67': 'y',
        '68': 'y',
        '69': 'y',
        '7': 'n',
        '70': 'y',
        '71': 'y',
        '72': 'y',
        '73': 'y',
        '74': 'y',
        '75': 'y',
        '76': 'y',
        '77': 'y',
        '78': 'y',
        '8': 'n',
        '9': 'n',
      },
      ie: {
        '10': 'n',
        '11': 'n',
        '6': 'n',
        '7': 'n',
        '8': 'n',
        '9': 'n',
      },
      ie_mob: {
        '10': 'n',
        '11': 'n',
      },
      ios_saf: {
        '10.0-10.2': 'n',
        '10.3': 'n',
        '11.0-11.2': 'n',
        '11.3-11.4': 'n',
        '12.0-12.1': 'n',
        '12.2-12.4': 'n',
        '13.0-13.1': 'n',
        '13.2': 'n',
        '13.3': 'n',
        '13.4': 'n',
        '3.2': 'n',
        '4.0-4.1': 'n',
        '4.2-4.3': 'n',
        '5.0-5.1': 'n',
        '6.0-6.1': 'n',
        '7.0-7.1': 'n',
        '8': 'n',
        '8.1-8.4': 'n',
        '9.0-9.2': 'n',
        '9.3': 'n',
      },
      kaios: {
        '2.5': 'n',
      },
      op_mini: {
        all: 'n',
      },
      op_mob: {
        '12': 'n',
        '12.1': 'n',
        '46': 'n',
      },
      opera: {
        '10.0-10.1': 'n',
        '11.5': 'n',
        '12.1': 'n',
        '15': 'n',
        '16': 'n',
        '17': 'n',
        '18': 'n',
        '19': 'n',
        '20': 'n',
        '21': 'n',
        '22': 'n',
        '23': 'n',
        '24': 'n',
        '25': 'n',
        '26': 'n',
        '27': 'n',
        '28': 'n',
        '29': 'n',
        '30': 'n',
        '31': 'n',
        '32': 'n',
        '33': 'n',
        '34': 'n',
        '35': 'n',
        '36': 'n',
        '37': 'n',
        '38': 'n',
        '39': 'n',
        '40': 'n',
        '41': 'n',
        '42': 'n',
        '43': 'n',
        '44': 'n',
        '45': 'n',
        '46': 'n',
        '47': 'n',
        '48': 'n',
        '49': 'n',
        '50': 'n',
        '51': 'n',
        '52': 'n',
        '53': 'n',
        '54': 'n',
        '55': 'n',
        '56': 'n',
        '57': 'n',
        '58': 'n',
        '60': 'n',
        '62': 'n',
        '63': 'n',
        '64': 'n',
        '65': 'n',
        '66': 'n',
        '67': 'n',
        '68': 'n',
      },
      safari: {
        '10': 'n',
        '10.1': 'n',
        '11': 'n',
        '11.1': 'n',
        '12': 'n',
        '12.1': 'n',
        '13': 'n',
        '13.1': 'n',
        '3.1': 'n',
        '3.2': 'n',
        '4': 'n',
        '5': 'n',
        '5.1': 'n',
        '6': 'n',
        '6.1': 'n',
        '7': 'n',
        '7.1': 'n',
        '8': 'n',
        '9': 'n',
        '9.1': 'n',
        TP: 'n',
      },
      samsung: {
        '10.1': 'n',
        '11.1': 'n',
        '4': 'n',
        '5.0-5.4': 'n',
        '6.2-6.4': 'n',
        '7.2-7.4': 'n',
        '8.2': 'n',
        '9.2': 'n',
      },
    },
    status: 'wd',
    ucprefix: false,
    webkit_id: '',
  },
  {
    baseCategories: ['CSS'],
    bug_count: 9,
    children: ['flexbox-gap'],
    chrome_id: '4837301406400512',
    date_published: '2009-07-23',
    description:
      'Method of positioning elements in horizontal or vertical stacks. Support includes all properties prefixed with `flex`, as well as `display: flex`, `display: inline-flex`, `align-content`, `align-items`, `align-self`, `justify-content` and `order`.',
    firefox_id: '',
    ie_id: 'flexbox',
    keywords:
      'flex-box,flex-direction,flex-wrap,flex-flow,flex-grow,flex-shrink,flex-basis,display:flex,justify-content:space-between,justify-content:space-around,align-content:space-between,align-content:space-around,space-between',
    link_count: 12,
    notes:
      'Most partial support refers to supporting an [older version](https://www.w3.org/TR/2009/WD-css3-flexbox-20090723/) of the specification or an [older syntax](https://www.w3.org/TR/2012/WD-css3-flexbox-20120322/).',
    notes_by_num: {
      '1':
        'Only supports the [old flexbox](https://www.w3.org/TR/2009/WD-css3-flexbox-20090723) specification and does not support wrapping.',
      '2':
        'Only supports the [2012 syntax](https://www.w3.org/TR/2012/WD-css3-flexbox-20120322/)',
      '3': 'Does not support flex-wrap, flex-flow or align-content properties',
      '4':
        'Partial support is due to large amount of bugs present (see known issues)',
    },
    parent: '',
    spec: 'https://www.w3.org/TR/css3-flexbox/',
    stats: {
      and_chr: {
        '81': 'y',
      },
      and_ff: {
        '68': 'y',
      },
      and_qq: {
        '10.4': 'y',
      },
      and_uc: {
        '12.12': 'y',
      },
      android: {
        '2.1': 'a x #1',
        '2.2': 'a x #1',
        '2.3': 'a x #1',
        '3': 'a x #1',
        '4': 'a x #1',
        '4.1': 'a x #1',
        '4.2-4.3': 'a x #1',
        '4.4': 'y',
        '4.4.3-4.4.4': 'y',
        '81': 'y',
      },
      baidu: {
        '7.12': 'y',
      },
      bb: {
        '10': 'y',
        '7': 'a x #1',
      },
      chrome: {
        '10': 'a x #1',
        '11': 'a x #1',
        '12': 'a x #1',
        '13': 'a x #1',
        '14': 'a x #1',
        '15': 'a x #1',
        '16': 'a x #1',
        '17': 'a x #1',
        '18': 'a x #1',
        '19': 'a x #1',
        '20': 'a x #1',
        '21': 'y x',
        '22': 'y x',
        '23': 'y x',
        '24': 'y x',
        '25': 'y x',
        '26': 'y x',
        '27': 'y x',
        '28': 'y x',
        '29': 'y',
        '30': 'y',
        '31': 'y',
        '32': 'y',
        '33': 'y',
        '34': 'y',
        '35': 'y',
        '36': 'y',
        '37': 'y',
        '38': 'y',
        '39': 'y',
        '4': 'a x #1',
        '40': 'y',
        '41': 'y',
        '42': 'y',
        '43': 'y',
        '44': 'y',
        '45': 'y',
        '46': 'y',
        '47': 'y',
        '48': 'y',
        '49': 'y',
        '5': 'a x #1',
        '50': 'y',
        '51': 'y',
        '52': 'y',
        '53': 'y',
        '54': 'y',
        '55': 'y',
        '56': 'y',
        '57': 'y',
        '58': 'y',
        '59': 'y',
        '6': 'a x #1',
        '60': 'y',
        '61': 'y',
        '62': 'y',
        '63': 'y',
        '64': 'y',
        '65': 'y',
        '66': 'y',
        '67': 'y',
        '68': 'y',
        '69': 'y',
        '7': 'a x #1',
        '70': 'y',
        '71': 'y',
        '72': 'y',
        '73': 'y',
        '74': 'y',
        '75': 'y',
        '76': 'y',
        '77': 'y',
        '78': 'y',
        '79': 'y',
        '8': 'a x #1',
        '80': 'y',
        '81': 'y',
        '83': 'y',
        '84': 'y',
        '85': 'y',
        '86': 'y',
        '9': 'a x #1',
      },
      edge: {
        '12': 'y',
        '13': 'y',
        '14': 'y',
        '15': 'y',
        '16': 'y',
        '17': 'y',
        '18': 'y',
        '79': 'y',
        '80': 'y',
        '81': 'y',
        '83': 'y',
      },
      firefox: {
        '10': 'a x #1',
        '11': 'a x #1',
        '12': 'a x #1',
        '13': 'a x #1',
        '14': 'a x #1',
        '15': 'a x #1',
        '16': 'a x #1',
        '17': 'a x #1',
        '18': 'a x #1',
        '19': 'a x #1',
        '2': 'a x #1',
        '20': 'a x #1',
        '21': 'a x #1',
        '22': 'a #3',
        '23': 'a #3',
        '24': 'a #3',
        '25': 'a #3',
        '26': 'a #3',
        '27': 'a #3',
        '28': 'y',
        '29': 'y',
        '3': 'a x #1',
        '3.5': 'a x #1',
        '3.6': 'a x #1',
        '30': 'y',
        '31': 'y',
        '32': 'y',
        '33': 'y',
        '34': 'y',
        '35': 'y',
        '36': 'y',
        '37': 'y',
        '38': 'y',
        '39': 'y',
        '4': 'a x #1',
        '40': 'y',
        '41': 'y',
        '42': 'y',
        '43': 'y',
        '44': 'y',
        '45': 'y',
        '46': 'y',
        '47': 'y',
        '48': 'y',
        '49': 'y',
        '5': 'a x #1',
        '50': 'y',
        '51': 'y',
        '52': 'y',
        '53': 'y',
        '54': 'y',
        '55': 'y',
        '56': 'y',
        '57': 'y',
        '58': 'y',
        '59': 'y',
        '6': 'a x #1',
        '60': 'y',
        '61': 'y',
        '62': 'y',
        '63': 'y',
        '64': 'y',
        '65': 'y',
        '66': 'y',
        '67': 'y',
        '68': 'y',
        '69': 'y',
        '7': 'a x #1',
        '70': 'y',
        '71': 'y',
        '72': 'y',
        '73': 'y',
        '74': 'y',
        '75': 'y',
        '76': 'y',
        '77': 'y',
        '78': 'y',
        '8': 'a x #1',
        '9': 'a x #1',
      },
      ie: {
        '10': 'a x #2 #4',
        '11': 'a #4',
        '6': 'n',
        '7': 'n',
        '8': 'n',
        '9': 'n',
      },
      ie_mob: {
        '10': 'a x #2',
        '11': 'y',
      },
      ios_saf: {
        '10.0-10.2': 'y',
        '10.3': 'y',
        '11.0-11.2': 'y',
        '11.3-11.4': 'y',
        '12.0-12.1': 'y',
        '12.2-12.4': 'y',
        '13.0-13.1': 'y',
        '13.2': 'y',
        '13.3': 'y',
        '13.4': 'y',
        '3.2': 'a x #1',
        '4.0-4.1': 'a x #1',
        '4.2-4.3': 'a x #1',
        '5.0-5.1': 'a x #1',
        '6.0-6.1': 'a x #1',
        '7.0-7.1': 'y x',
        '8': 'y x',
        '8.1-8.4': 'y x',
        '9.0-9.2': 'y',
        '9.3': 'y',
      },
      kaios: {
        '2.5': 'y',
      },
      op_mini: {
        all: 'y',
      },
      op_mob: {
        '12': 'n',
        '12.1': 'y',
        '46': 'y',
      },
      opera: {
        '10.0-10.1': 'n',
        '11.5': 'n',
        '12.1': 'y',
        '15': 'y x',
        '16': 'y x',
        '17': 'y',
        '18': 'y',
        '19': 'y',
        '20': 'y',
        '21': 'y',
        '22': 'y',
        '23': 'y',
        '24': 'y',
        '25': 'y',
        '26': 'y',
        '27': 'y',
        '28': 'y',
        '29': 'y',
        '30': 'y',
        '31': 'y',
        '32': 'y',
        '33': 'y',
        '34': 'y',
        '35': 'y',
        '36': 'y',
        '37': 'y',
        '38': 'y',
        '39': 'y',
        '40': 'y',
        '41': 'y',
        '42': 'y',
        '43': 'y',
        '44': 'y',
        '45': 'y',
        '46': 'y',
        '47': 'y',
        '48': 'y',
        '49': 'y',
        '50': 'y',
        '51': 'y',
        '52': 'y',
        '53': 'y',
        '54': 'y',
        '55': 'y',
        '56': 'y',
        '57': 'y',
        '58': 'y',
        '60': 'y',
        '62': 'y',
        '63': 'y',
        '64': 'y',
        '65': 'y',
        '66': 'y',
        '67': 'y',
        '68': 'y',
      },
      safari: {
        '10': 'y',
        '10.1': 'y',
        '11': 'y',
        '11.1': 'y',
        '12': 'y',
        '12.1': 'y',
        '13': 'y',
        '13.1': 'y',
        '3.1': 'a x #1',
        '3.2': 'a x #1',
        '4': 'a x #1',
        '5': 'a x #1',
        '5.1': 'a x #1',
        '6': 'a x #1',
        '6.1': 'y x',
        '7': 'y x',
        '7.1': 'y x',
        '8': 'y x',
        '9': 'y',
        '9.1': 'y',
        TP: 'y',
      },
      samsung: {
        '10.1': 'y',
        '11.1': 'y',
        '4': 'y',
        '5.0-5.4': 'y',
        '6.2-6.4': 'y',
        '7.2-7.4': 'y',
        '8.2': 'y',
        '9.2': 'y',
      },
    },
    status: 'cr',
    ucprefix: false,
    webkit_id: '',
  },
];
