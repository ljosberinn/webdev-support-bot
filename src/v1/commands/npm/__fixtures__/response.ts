export const response = [
  {
    date: '2020-03-19T19:53:13.309Z',
    description: 'React is a JavaScript library for building user interfaces.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
    ],
    name: 'react',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Jed Watson',
    },
    date: '2020-03-22T04:02:30.633Z',
    description: 'A Select control built with and for ReactJS',
    keywords: [
      'combobox',
      'form',
      'input',
      'multiselect',
      'react',
      'react-component',
      'select',
      'ui',
    ],
    links: {
      bugs: 'https://github.com/JedWatson/react-select/issues',
      homepage: 'https://github.com/JedWatson/react-select#readme',
      npm: 'https://www.npmjs.com/package/react-select',
      repository: 'https://github.com/JedWatson/react-select',
    },
    maintainers: [
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'cc.lee@live.com.au',
        username: 'gwyneplaine',
      },
      {
        email: 'jed.watson@me.com',
        username: 'jedwatson',
      },
      {
        email: 'mitchell@mitchellhamilton.me',
        username: 'mitchellhamilton',
      },
      {
        email: 'automation+react-select@thinkmill.com.au',
        username: 'react-select-release-bot',
      },
    ],
    name: 'react-select',
    publisher: {
      email: 'automation+react-select@thinkmill.com.au',
      username: 'react-select-release-bot',
    },
    scope: 'unscoped',
    version: '3.1.0',
  },
  {
    author: {
      name: 'Matt Zabriskie',
    },
    date: '2020-05-14T14:38:16.281Z',
    description: 'React draggable component',
    keywords: ['react', 'draggable', 'react-component'],
    links: {
      bugs: 'https://github.com/mzabriskie/react-draggable/issues',
      homepage: 'https://github.com/mzabriskie/react-draggable',
      npm: 'https://www.npmjs.com/package/react-draggable',
      repository: 'https://github.com/mzabriskie/react-draggable',
    },
    maintainers: [
      {
        email: 'mzabriskie@gmail.com',
        username: 'mzabriskie',
      },
      {
        email: 'samuel.trace.reed@gmail.com',
        username: 'strml',
      },
    ],
    name: 'react-draggable',
    publisher: {
      email: 'samuel.trace.reed@gmail.com',
      username: 'strml',
    },
    scope: 'unscoped',
    version: '4.4.2',
  },
  {
    author: {
      email: 'travis@souporserious.com',
      name: 'Travis Arnold',
      url: 'http://souporserious.com',
    },
    date: '2020-04-28T09:55:08.691Z',
    description: 'Official library to use Popper on React projects',
    keywords: [
      'react',
      'react-popper',
      'popperjs',
      'component',
      'drop',
      'tooltip',
      'popover',
    ],
    links: {
      bugs: 'https://github.com/popperjs/react-popper/issues',
      homepage: 'https://github.com/popperjs/react-popper',
      npm: 'https://www.npmjs.com/package/react-popper',
      repository: 'https://github.com/popperjs/react-popper',
    },
    maintainers: [
      {
        email: 'federico.zivolo@gmail.com',
        username: 'fezvrasta',
      },
      {
        email: 'ftntravis@gmail.com',
        username: 'souporserious',
      },
    ],
    name: 'react-popper',
    publisher: {
      email: 'federico.zivolo@gmail.com',
      username: 'fezvrasta',
    },
    scope: 'unscoped',
    version: '2.2.3',
  },
  {
    author: {
      name: 'Param Aggarwal',
    },
    date: '2020-04-26T08:29:08.419Z',
    description: 'Simple HTML5 drag-drop zone with React.js',
    keywords: ['react-component', 'react', 'drag', 'drop', 'upload'],
    links: {
      bugs: 'https://github.com/react-dropzone/react-dropzone/issues',
      homepage: 'https://github.com/react-dropzone/react-dropzone',
      npm: 'https://www.npmjs.com/package/react-dropzone',
      repository: 'https://github.com/react-dropzone/react-dropzone',
    },
    maintainers: [
      {
        email: 'andrey@okonet.ru',
        username: 'okonet',
      },
      {
        email: 'rolandjitsu@gmail.com',
        username: 'rolandjitsu',
      },
      {
        email: 'rmarkins@gmail.com',
        username: 'rxmarbles',
      },
    ],
    name: 'react-dropzone',
    publisher: {
      email: 'rolandjitsu@gmail.com',
      username: 'rolandjitsu',
    },
    scope: 'unscoped',
    version: '11.0.1',
  },
  {
    author: {
      email: 'dan.abramov@me.com',
      name: 'Dan Abramov',
      url: 'https://github.com/gaearon',
    },
    date: '2020-02-18T03:43:52.869Z',
    description: 'Official React bindings for Redux',
    keywords: ['react', 'reactjs', 'redux'],
    links: {
      bugs: 'https://github.com/reduxjs/react-redux/issues',
      homepage: 'https://github.com/reduxjs/react-redux',
      npm: 'https://www.npmjs.com/package/react-redux',
      repository: 'https://github.com/reduxjs/react-redux',
    },
    maintainers: [
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'timdorr@timdorr.com',
        username: 'timdorr',
      },
    ],
    name: 'react-redux',
    publisher: {
      email: 'mark.erikson@gmail.com',
      username: 'acemarke',
    },
    scope: 'unscoped',
    version: '7.2.0',
  },
  {
    author: {
      email: 'case@casesandberg.com',
      name: 'case',
      username: 'case',
    },
    date: '2020-05-04T00:55:19.458Z',
    description:
      'A Collection of Color Pickers from Sketch, Photoshop, Chrome & more',
    keywords: [
      'react',
      'color picker',
      'react-component',
      'colorpicker',
      'picker',
      'sketch',
      'chrome',
      'photoshop',
      'material design',
      'popup',
    ],
    links: {
      homepage: 'http://casesandberg.github.io/react-color/',
      npm: 'https://www.npmjs.com/package/react-color',
    },
    maintainers: [
      {
        email: 'case@casesandberg.com',
        username: 'case',
      },
    ],
    name: 'react-color',
    publisher: {
      email: 'case@casesandberg.com',
      username: 'case',
    },
    scope: 'unscoped',
    version: '2.18.1',
  },
  {
    date: '2020-03-19T19:53:23.004Z',
    description: 'React package for working with the DOM.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react-dom',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
      {
        email: 'paul@oshannessy.com',
        username: 'zpao',
      },
    ],
    name: 'react-dom',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Felix Kling',
    },
    date: '2020-02-26T00:42:09.515Z',
    description:
      'A CLI and toolkit to extract information from React components for documentation generation.',
    keywords: ['react', 'documentation-generation'],
    links: {
      bugs: 'https://github.com/reactjs/react-docgen/issues',
      homepage: 'https://github.com/reactjs/react-docgen#readme',
      npm: 'https://www.npmjs.com/package/react-docgen',
      repository: 'https://github.com/reactjs/react-docgen',
    },
    maintainers: [
      {
        email: 'daniel@tschinder.de',
        username: 'danez',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
    ],
    name: 'react-docgen',
    publisher: {
      email: 'daniel@tschinder.de',
      username: 'danez',
    },
    scope: 'unscoped',
    version: '5.3.0',
  },
  {
    date: '2020-05-15T21:38:04.997Z',
    description:
      'Hooks for building lightweight, fast and extendable datagrids for React',
    keywords: ['react', 'table', 'react-table', 'datagrid'],
    links: {
      bugs: 'https://github.com/tannerlinsley/react-table/issues',
      homepage: 'https://react-table.js.org',
      npm: 'https://www.npmjs.com/package/react-table',
      repository: 'https://github.com/tannerlinsley/react-table',
    },
    maintainers: [
      {
        email: 'tannerlinsley@gmail.com',
        username: 'tannerlinsley',
      },
      {
        email: 'zach.h.sherman@gmail.com',
        username: 'zsherman',
      },
    ],
    name: 'react-table',
    publisher: {
      email: 'tannerlinsley@gmail.com',
      username: 'tannerlinsley',
    },
    scope: 'unscoped',
    version: '7.1.0',
  },
];
