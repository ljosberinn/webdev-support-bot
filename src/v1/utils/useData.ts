import * as Sentry from '@sentry/node';
import fetch, {
  HeaderInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env';
import { Cache } from './Cache';

const apiCache = new Cache({
  checkperiod: Number.parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  maxKeys: Number.parseInt(API_CACHE_ENTRIES_LIMIT, 10),
  stdTTL: Number.parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper) => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get<TParsedResponse>(casedCacheKey);

  if (cachedResponse) {
    return () => cachedResponse;
  }

  return async (url, fetchOptions) => {
    Sentry.addBreadcrumb({
      category: 'query',
      data: typeof url === 'string' ? { url } : undefined,
      level: Sentry.Severity.Info,
      timestamp: Date.now(),
    });

    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    // eslint-disable-next-line no-console
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    // eslint-disable-next-line no-console
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = (
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    };
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    };
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    };
  }
};

const useData = <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeaderInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};

export default useData;
