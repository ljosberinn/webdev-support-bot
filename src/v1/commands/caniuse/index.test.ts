import { getChosenResult } from '../../utils/discordTools';
import { getData } from '../../utils/urlTools';
import useData from '../../utils/useData';
import { simpleResponse, detailedResponse } from './__fixtures__/response';

import { buildCanIUseQueryHandler } from '.';

describe('caniuse', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const fetchDetails: jest.MockedFunction<typeof useData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  beforeEach(() => {
    fetch.mockResolvedValue(simpleResponse);
    fetchDetails.mockResolvedValue({
      error: false,
      json: detailedResponse,
      text: null,
    });
  });

  afterEach(() => jest.resetAllMocks());

  test('returns when request fails', async () => {
    fetch.mockResolvedValueOnce(null);
    const handler = buildCanIUseQueryHandler(fetch);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'caniuse',
      sanitizeData: expect.any(Function),
      searchTerm: 'search term',
    });

    expect(choose).not.toBeCalled();
  });

  test('awaits user response when request works', async () => {
    const editMock = jest.fn();
    choose.mockResolvedValue({
      author: { name: '', url: '' },
      description: '',
      externalUrls: { homepage: '', repository: '' },
      lastUpdate: '',
      maintainers: '10',
      name: 'React',
      url: 'http://react',
    });

    msg.channel.send.mockResolvedValue({ edit: editMock });

    const handler = buildCanIUseQueryHandler(fetch, fetchDetails, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'caniuse',
      sanitizeData: expect.any(Function),
      searchTerm: 'search term',
    });

    expect(msg.channel.send.mock.calls[0]).toMatchSnapshot();
    expect(choose.mock.calls[0]).toMatchSnapshot();
    expect(sendMock.mock.calls[0]).toMatchSnapshot();
  });
});
