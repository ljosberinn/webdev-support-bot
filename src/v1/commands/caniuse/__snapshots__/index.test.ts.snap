// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`caniuse awaits user response when request works 1`] = `undefined`;

exports[`caniuse awaits user response when request works 2`] = `undefined`;

exports[`caniuse awaits user response when request works 3`] = `undefined`;
